language/src/rustspec_to_coq_base.rs

```diff
@@ -314,53 +316,68 @@ pub(crate) fn array_or_seq<'a>(t: Typ, top_ctxt: &'a TopLevelContext) -> RcDoc<'
+                                                },
+                                                DUMMY_SP.into(),
+                                            ),
+                                            Some(vec![t_base.clone()]),
+                                        ),
+                                        "Some",
                                     ),
-                                    None,
-                                ),
-                                (
-                                    TopLevelIdent {
-                                        string: match carrier_kind(ert) {
-                                            EarlyReturnType::Option => "Some",
-                                            EarlyReturnType::Result => "Ok",
-                                        }
-                                        .to_string(),
-                                        kind: TopLevelIdentKind::EnumConstructor,
-                                    },
-                                    DUMMY_SP.into(),
-                                ),
-                                Some((Box::new(e.clone()), DUMMY_SP.into())),
-                            ),
-                            t,
-                        ),
-                        _ => panic!("should not happen"),
-                    }
+                                    CarrierTyp::Result(a, b) => (
+                                        BaseTyp::Named(
+                                            (
+                                                TopLevelIdent {
+                                                    string: "Result".to_string(),
+                                                    kind: TopLevelIdentKind::Type,
+                                                },
+                                                DUMMY_SP.into(),
+                                            ),
+                                            Some(vec![t_base.clone(), b]),
+                                        ),
+                                        "Ok",
+                                    ),
+                                };
+
+                                Statement::ReturnExp(
+                                    Expression::EnumInject(
+                                        early_typ.clone(),
+                                        (
+                                            TopLevelIdent {
+                                                string: fun_name.to_string(),
+                                                kind: TopLevelIdentKind::EnumConstructor,
+                                            },
+                                            DUMMY_SP.into(),
+                                        ),
+                                        Some((Box::new(e.clone()), DUMMY_SP.into())),
+                                    ),
+                                    Some((x, (early_typ.clone(), t_base.1))),
+                                )
```

Comment:
I think you could replace that chunk of code with something like `pure_carrier(ert, e)` here

Reply:
It seems the carrier has an incorrect type, e.g. I have to use `t_base` instead of `a` in args.

Reply:
I can do it as
```rust
match stmt {
    Statement::ReturnExp(e, Some((x, t_base))) => {
        let carrier = match ert.clone() {
            CarrierTyp::Option(_) => CarrierTyp::Option(t_base.clone()),
            CarrierTyp::Result(_, b) => CarrierTyp::Result(t_base.clone(), b),
        };
        let (e, _) = pure_carrier(carrier.clone(), (e.clone(), DUMMY_SP.into()));
        Statement::ReturnExp(e, Some((x, (carrier.clone().into(), t_base.1))))
    }
    _ => panic!("should not happen"),
}

```